dist/screeps-tests.ts

```diff
@@ -1030,8 +1030,165 @@ function resources(o: GenericStore): ResourceConstant[] {
+            T3: "XUH2O",
+        },
+        carry: {
+            T1: "KH",
+            T2: "KH2O",
+            T3: "XKH2O",
+        },
+        ranged_attack: {
+            T1: "KO",
+            T2: "KHO2",
+            T3: "XKHO2",
+        },
+        heal: {
+            T1: "LO",
+            T2: "LHO2",
+            T3: "XLHO2",
+        },
+        move: {
+            T1: "ZO",
+            T2: "ZHO2",
+            T3: "XZHO2",
+        },
+        tough: {
+            T1: "GO",
+            T2: "GHO2",
+            T3: "XGHO2",
+        },
+        harvest: {
+            T1: "UO",
+            T2: "UHO2",
+            T3: "XUHO2",
+        },
+        construct: {
+            T1: "LH",
+            T2: "LH2O",
+            T3: "XLH2O",
+        },
+        dismantle: {
+            T1: "ZH",
+            T2: "ZH2O",
+            T3: "XZH2O",
+        },
+        upgrade: {
+            T1: "GH",
+            T2: "GH2O",
+            T3: "XGH2O",
+        },
+    };
+
+    const BoostTypeBodyparts: {
+        [boostType in BoostType]: BodyPartConstant;
+    } = {
+        [ATTACK]: ATTACK,
+        [CARRY]: CARRY,
+        [RANGED_ATTACK]: RANGED_ATTACK,
+        [HEAL]: HEAL,
+        [MOVE]: MOVE,
+        [TOUGH]: TOUGH,
+        [HARVEST]: WORK,
+        [CONSTRUCT]: WORK,
+        [DISMANTLE]: WORK,
+        [UPGRADE]: WORK,
+    };
+
+    const BoostTypeToBoostArray: {
+        [boostType in BoostType]: BoostModifier;
+    } = {
+        [ATTACK]: ATTACK,
+        [CARRY]: "capacity",
+        [RANGED_ATTACK]: "rangedAttack",
+        // [RANGED_MASS_ATTACK]: "rangedMassAttack",
+        [HEAL]: HEAL,
+        [MOVE]: "fatigue",
+        [TOUGH]: "damage",
+        [HARVEST]: "harvest",
+        [CONSTRUCT]: "build",
+        // [REPAIR]: "repair",
+        [DISMANTLE]: "dismantle",
+        [UPGRADE]: "upgradeController",
+    };
+
+    /**
+     *
+     * @param body
+     * @param type
+     * @param intendedBoosts
+     * @returns
+     */
+    function getBodyPotential(body: BodyPartDefinition[], type: BoostType, intendedBoosts: MineralBoostConstant[] = []): number {
+        const bodyPart = BoostTypeBodyparts[type];
+        return body.reduce((sum, part) => {
+            if (part.hits === 0) {
+                return sum + 0;
+            }
+            if (part.type === bodyPart) {
+                let boost = part.boost;
+                if (!boost && intendedBoosts) {
+                    boost = intendedBoosts.find(
+                        (boost) => boost === BOOST_TIERS[type].T1 || boost === BOOST_TIERS[type].T2 || boost === BOOST_TIERS[type].T3,
+                    );
+                }
+                if (!boost) {
+                    return sum + 1;
+                }
+
+                const key = BoostTypeToBoostArray[type];
+                const partBoost = BOOSTS[bodyPart];
+                if (!partBoost || !(boost in partBoost) || !partBoost[boost] || !(key in partBoost[boost])) {
```

Comment:
All of the `partBoost[boost]` checks error with
```
Element implicitly has an 'any' type because expression of type 'string | number' can't be used to index type '{ [boost: string]: { [action: string]: number; }; } | { UO: { harvest: 3; }; UHO2: { harvest: 5; }; XUHO2: { harvest: 7; }; LH: { build: 1.5; repair: 1.5; }; LH2O: { build: 1.8; repair: 1.8; }; XLH2O: { build: 2; repair: 2; }; ... 5 more ...; XGH2O: { ...; }; } | ... 5 more ... | { ...; }'.
  No index signature with a parameter of type 'string' was found on type '{ [boost: string]: { [action: string]: number; }; } | { UO: { harvest: 3; }; UHO2: { harvest: 5; }; XUHO2: { harvest: 7; }; LH: { build: 1.5; repair: 1.5; }; LH2O: { build: 1.8; repair: 1.8; }; XLH2O: { build: 2; repair: 2; }; ... 5 more ...; XGH2O: { ...; }; } | ... 5 more ... | { ...; }'.ts(7053)
```